interfaces/polkit/backend.go

```diff
@@ -0,0 +1,116 @@
+ */
+
+// Package polkit implements interaction between snapd and polkit.
+//
+// Snapd installs polkitd policy files on behalf of snaps that
+// describe administrative actions they can perform on behalf of
+// clients.
+//
+// The policy files are XML files whose format is described here:
+// https://www.freedesktop.org/software/polkit/docs/latest/polkit.8.html#polkit-declaring-actions
+package polkit
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/timings"
+)
+
+// Backend is responsible for maintaining DBus policy files.
+type Backend struct{}
+
+// Initialize does nothing.
+func (b *Backend) Initialize(*interfaces.SecurityBackendOptions) error {
+	return nil
+}
+
+// Name returns the name of the backend.
+func (b *Backend) Name() interfaces.SecuritySystem {
+	return interfaces.SecurityPolkit
+}
+
+// Setup installs the polkit policy files specific to a given snap.
+//
+// Polkit has no concept of a complain mode so confinment type is ignored.
+func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions, repo *interfaces.Repository, tm timings.Measurer) error {
+	snapName := snapInfo.InstanceName()
+	// Get the policies that apply to this snap
+	spec, err := repo.SnapSpecification(b.Name(), snapName)
+	if err != nil {
+		return fmt.Errorf("cannot obtain polkit specification for snap %q: %s", snapName, err)
+	}
+
+	// Get the files that this snap should have
+	glob := interfaces.InterfacePolkitPolicyName(snapName, "*")
+	content := deriveContent(spec.(*Specification), snapInfo)
+	dir := dirs.SnapPolkitPolicyDir
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for polkit policy files %q: %s", dir, err)
+	}
+	_, _, err = osutil.EnsureDirState(dir, glob, content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize polkit policy files for snap %q: %s", snapName, err)
+	}
+	return nil
+}
+
+// Remove removes dbus configuration files of a given snap.
+//
+// This method should be called after removing a snap.
+func (b *Backend) Remove(snapName string) error {
+	glob := interfaces.InterfacePolkitPolicyName(snapName, "*")
+	_, _, err := osutil.EnsureDirState(dirs.SnapPolkitPolicyDir, glob, nil)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize polkit files for snap %q: %s", snapName, err)
+	}
+	return nil
+}
+
+// deriveContent combines security snippets collected from all the interfaces
+// affecting a given snap into a content map applicable to EnsureDirState.
+func deriveContent(spec *Specification, snapInfo *snap.Info) map[string]osutil.FileState {
+	policies := spec.Policies()
+	if len(policies) == 0 {
+		return nil
+	}
+	content := make(map[string]osutil.FileState, len(policies)+1)
+	for name, policyContent := range policies {
+		filename := interfaces.InterfacePolkitPolicyName(snapInfo.InstanceName(), name)
```

Comment:
apparently the systemd backend has the interfaces themselves call this instead. I don't have a strong preference but is a bit strange/confusing to have both approaches in use.

Reply:
I created as a draft https://github.com/snapcore/snapd/pull/10516  which changes the systemd backend to do this as well, it also introduces some conventions in terms of naming/doc comments around this.  Once there's  a bit of an agreement on that one this one should follow suit and be tweaked to follow the new patterns there.

Reply:
that PR is now landed and this one can be tweaked